README.md
# Roger-Q

[![Latest Version](https://img.shields.io/github/release/liip/roger-q.svg?style=flat-square)](https://gitreleases.dev/gh/liip/roger-q/latest/roger-q.phar)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Build Status](https://img.shields.io/travis/liip/roger-q/master.svg?style=flat-square)](https://travis-ci.org/liip/roger-q)

Roger-Q is a tool to handle RabbitMQ queues, and it includes commands to dump, dedupe and publish messages.

## Installation

Run this to download the latest release from Github, move it into your bin folder, and make it executable:
```bash
curl -L https://github.com/liip/roger-q/releases/latest/download/roger-q.phar > /usr/local/bin/roger-q
chmod a+x /usr/local/bin/roger-q
```

## Commands

To get the list of all commands, run `roger-q.phar list`

To see all options of a command, run `roger-q.phar {commmand} --help`.

Except for dedupe, the commands require the `--host` option to know where to find RabbitMQ, and `--username`,
`--password` and `--vhost`. The default is localhost on port 5672 with vhost `/` and credentials guest/guest.
(Hint: Pull request to support a configuration file for these parameters would be welcome.)

### Dump (`dump`)

Dumps the messages from the given queue to the standard output. The command uses the RabbitMQ management API, rather
than the regular protocol, to read and write messages, which is much faster for queues with many messages. This means
that you need to install the [Management Plugin](https://www.rabbitmq.com/management.html) in your RabbitMQ. If you do
not use the default port 15672, you need to specify the `--port` parameter.

The messages are dumped as JSON as in the following example (the example has been pretty-formatted):
```json5
[
  {
     "payload_bytes": 129,
     "redelivered": false,
     "exchange": "my-exchange-name",
     "routing_key": "routing-key",
     "message_count": 3891,
     "properties": {
       // Message headers here
     },
     "payload": "The message payload",
     "payload_encoding": "string"
   },
   // Next messages here...
]
```

Example usage:

```bash
# roger-q.phar dump queue-name
```

### Dedupe (`dedupe`)

In the following example the messages from `messages.txt` file will be de-duplicated by checking the uniqueness of the
given field.
The resulting messages are outputted to the standard output.

```bash
# cat messages.txt | roger-q.phar dedupe id-field
```

### Publish (`publish`)

Reads messages from the standard input and publishes them to the specified *queue*. Usually, you would send messages to
an exchange rather than a queue directly, but the use-case of this tool is to clean up queues. If you specify an
exchange instead of a queue, the messages will be reported as published but not arrive anywhere.

The following example will publish the messages from the `messages.txt` file into the `queue-name` queue.
The queue is additionally purged before starting the publishing operation.

```bash
# cat messages.txt | roger-q.phar publish queue-name --purge
```

## PHAR building

[Box](https://github.com/humbug/box) is used to build a precompiled PHAR file of the application.

Run `make dist` to:

-   download the box library in `tools/` (if not present)
-   create the PHAR executable in `dist/roger-q.phar`

## Development

To run the PHP coding-styles checks (`php-cs-fixer` and `phpstan`) run the `make phpcs` command to:

-   download the `php-cs-fixer` tool in `tools/` (if not present)
-   download the `phpstan` tool in `tools/` (if not present)
-   Run `php-cs-fixer` on the source code
-   Run `phpstan` on the source code


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

## 0.1.0

- Updated depencencies

## 0.0.2
### Added
- Support de-duplication on multiple payload fields [PR#8](https://github.com/liip/roger-q/pull/8) by [@thePanz](https://github.com/thePanz)
### Changed
- Using pnz/json-exception for JSON decode/encode exception handling
### Fixed
- Use correct default port and catch port in hostname [PR#6](https://github.com/liip/roger-q/pull/6) by [@dbu](https://github.com/dbu)
- Stopped using RequestOptions::SINK because it caused stream seek errors on some systems [PR#19](https://github.com/liip/roger-q/pull/19) by [@dbu](https://github.com/dbu)

## 0.0.1
### Fixed
- Fixed TravisCI deployment process and phar releasing

## 0.0.1-beta.1
### Added
- Added TravisCI integration
- Added documentation
- Added commands: dump, dedupe, publish


src/Command/Dump.php
<?php

declare(strict_types=1);

namespace App\Command;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Uri;
use GuzzleHttp\RequestOptions;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\ConsoleOutputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Dump extends Command
{
    /**
     * We dump from the HTTP API which runs on a different port than amqp.
     */
    private const DEFAULT_RABBITMQ_PORT = 15672;

    protected static $defaultName = 'dump';

    protected function configure(): void
    {
        $this
            ->setDescription('Dumps the messages of the specified queue to the standard output')
            ->addOption('host', null, InputOption::VALUE_REQUIRED, 'RabbitMQ host to connect to', 'localhost')
            ->addOption('port', null, InputOption::VALUE_REQUIRED, 'RabbitMQ HTTP API port', self::DEFAULT_RABBITMQ_PORT)
            ->addOption('username', null, InputOption::VALUE_REQUIRED, 'Username for the RabbitMQ connection', 'guest')
            ->addOption('password', null, InputOption::VALUE_REQUIRED, 'Password for the RabbitMQ connection', 'guest')
            ->addOption('vhost', null, InputOption::VALUE_REQUIRED, 'RabbitMQ VHost where the queue is declared', '/')
            ->addOption('limit', null, InputOption::VALUE_REQUIRED, 'Number of messages to dump')
            ->addOption('ackmode', null, InputOption::VALUE_REQUIRED, 'Determines whether the messages will be removed from the queue. Valid options are: "reject_requeue_true", "ack_requeue_true", "ack_requeue_false", "reject_requeue_false"', 'reject_requeue_true')
            ->addArgument('queue', InputArgument::REQUIRED, 'Name of the queue to dump')
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $queueName = $input->getArgument('queue');
        $host = $input->getOption('host');
        $vHost = $input->getOption('vhost');
        \assert(\is_string($queueName));
        \assert(\is_string($host));
        \assert(\is_string($vHost));

        $uri = new Uri('http://'.$host);
        if ($uri->getPort()) {
            throw new \UnexpectedValueException('You can not specify the port as part of the hostname. Use the separate "port" option.');
        }
        $uri = $uri->withPort($input->getOption('port'));

        $guzzle = new Client([
            'base_uri' => $uri,
            RequestOptions::CONNECT_TIMEOUT => 0,
            RequestOptions::READ_TIMEOUT => 0,
            RequestOptions::TIMEOUT => 0,
            RequestOptions::AUTH => [
                $input->getOption('username'),
                $input->getOption('password'),
            ],
        ]);

        $data = [
            'count' => $input->getOption('limit'),
            'ackmode' => $input->getOption('ackmode'),
            'encoding' => 'auto',
        ];
        if (null === $data['count']) {
            $data['count'] = \PHP_INT_MAX;
        }

        $response = $guzzle->request('POST', sprintf('/api/queues/%s/%s/get', $vHost, $queueName), [
            RequestOptions::HEADERS => [
                'Accept-Encoding' => 'gzip',
                'Transfer-Encoding' => 'chunked',
            ],
            RequestOptions::JSON => $data,
        ]);
        if (200 !== $response->getStatusCode()) {
            throw new \RuntimeException('Error '.$response->getStatusCode()."\n".$response->getBody()->getContents());
        }

        $stream = $response->getBody();
        while (!$stream->eof()) {
            $output->write($stream->read(100000));
        }

        if ($output instanceof ConsoleOutputInterface && $output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE) {
            $output->getErrorOutput()->writeln(sprintf(
                'Dumped <info>%s</info> bytes (gzip) from queue <info>%s</info>',
                $response->hasHeader('x-encoded-content-length') ? $response->getHeaderLine('x-encoded-content-length') : 'unspecified',
                $queueName
            ));
        }

        return Command::SUCCESS;
    }
}


src/Command/Dedupe.php
<?php

declare(strict_types=1);

namespace App\Command;

use Pnz\JsonException\Json;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\ConsoleOutputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Dedupe extends Command
{
    protected static $defaultName = 'dedupe';

    protected function configure(): void
    {
        $this
            ->setDescription('Removes duplicated messages')
            ->addArgument('field', InputArgument::REQUIRED | InputArgument::IS_ARRAY, 'Name of the JSON payload fields which identifies each message')
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        /** @var string[] $fields */
        $fields = $input->getArgument('field');
        \assert(\is_array($fields));

        $data = $this->readStdin();
        $messages = Json::decode($data, true);

        $removedMessagesCount = 0;
        $seenValues = [];
        $cleanMessages = [];
        foreach ($messages as $i => $message) {
            if ($this->isDuplicatedMessage($fields, $message, $seenValues, $i)) {
                ++$removedMessagesCount;
            } else {
                $cleanMessages[] = $message;
            }
        }

        $output->writeln(Json::encode($cleanMessages));

        if ($output instanceof ConsoleOutputInterface && $output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE) {
            $output->getErrorOutput()->writeln(sprintf(
                'Removed <info>%d</info> duplicated messages out of <info>%d</info> total messages, resulting in <info>%d</info> messages',
                $removedMessagesCount,
                \count($messages),
                \count($cleanMessages)
            ));
        }

        return Command::SUCCESS;
    }

    /**
     * @param string[] $fields
     * @param mixed[]  $data
     */
    private function getValuesHash(array $fields, array $data): string
    {
        $values = [];
        foreach ($fields as $field) {
            $values[$field] = $data[$field];
        }

        return sha1(serialize($values));
    }

    /**
     * @param string[]            $fields
     * @param mixed[]             $message
     * @param array<string, bool> $seenValues Hashmap of value-hash => bool, to keep track of values already encountered
     */
    private function isDuplicatedMessage(array $fields, array $message, array &$seenValues, int $messageNum): bool
    {
        if (!\array_key_exists('payload', $message)) {
            throw new \UnexpectedValueException(sprintf('Message #%d does not have a payload (%s)', $messageNum, Json::encode($message)));
        }

        $payload = Json::decode($message['payload'], true);

        // Check that all given fields exist in the payload:
        foreach ($fields as $field) {
            if (!\array_key_exists($field, $payload)) {
                throw new \UnexpectedValueException(sprintf('Payload of message #%d does not have the required fields %s (%s)', $messageNum, implode(',', $fields), $message['payload']));
            }
        }

        $hash = $this->getValuesHash($fields, $payload);
        if (\array_key_exists($hash, $seenValues)) {
            return true;
        }

        $seenValues[$hash] = true;

        return false;
    }

    private function readStdin(): string
    {
        $data = '';
        while (!feof(\STDIN)) {
            $data .= fread(\STDIN, 1024);
        }

        if ('' === $data) {
            throw new \InvalidArgumentException('No data received from STDIN');
        }

        return $data;
    }
}


src/Command/Publish.php
<?php

declare(strict_types=1);

namespace App\Command;

use Enqueue\AmqpBunny\AmqpConnectionFactory;
use Interop\Amqp\Impl\AmqpMessage;
use Pnz\JsonException\Json;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\ConsoleOutputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class Publish extends Command
{
    /**
     * We publish with the amqp protocol.
     */
    private const DEFAULT_RABBITMQ_PORT = 5672;

    protected static $defaultName = 'publish';

    protected function configure(): void
    {
        $this
            ->setDescription('Publishes messages from STDIN to the specified queue (NOT an exchange but only a queue)')
            ->addOption('host', null, InputOption::VALUE_REQUIRED, 'RabbitMQ host to connect to', 'localhost')
            ->addOption('port', null, InputOption::VALUE_REQUIRED, 'RabbitMQ message port', self::DEFAULT_RABBITMQ_PORT)
            ->addOption('username', null, InputOption::VALUE_REQUIRED, 'Username for the RabbitMQ connection', 'guest')
            ->addOption('password', null, InputOption::VALUE_REQUIRED, 'Password for the RabbitMQ connection', 'guest')
            ->addOption('vhost', null, InputOption::VALUE_REQUIRED, 'RabbitMQ VHost where the queue is declared', '/')
            ->addOption('purge', null, InputOption::VALUE_NONE, 'Whether or not the queue should be purged before publishing')
            ->addArgument('queue', InputArgument::REQUIRED, 'Name of the queue to publish the messages to')
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $host = $input->getOption('host');
        \assert(\is_string($host), 'host name is a string');
        if (false !== mb_strpos($host, ':')) {
            throw new \UnexpectedValueException('You can not specify the port as part of the hostname. Use the separate "port" option.');
        }

        $queueName = $input->getArgument('queue');
        \assert(\is_string($queueName));

        $data = $this->readStdin();
        $messages = Json::decode($data, true);

        $factory = new AmqpConnectionFactory([
            'host' => $input->getOption('host'),
            'port' => $input->getOption('port'),
            'vhost' => $input->getOption('vhost'),
            'user' => $input->getOption('username'),
            'pass' => $input->getOption('password'),
            'persisted' => false,
        ]);

        $context = $factory->createContext();
        $queue = $context->createQueue($queueName);
        $producer = $context->createProducer();

        if ($input->getOption('purge')) {
            $context->purgeQueue($queue);
        }

        foreach ($messages as $message) {
            $msg = new AmqpMessage($message['payload'] ?? '', [], $message['properties'] ?? []);
            $msg->setRoutingKey($message['routing_key'] ?? null);
            $producer->send($queue, $msg);
        }

        if ($output instanceof ConsoleOutputInterface && $output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE) {
            $output->getErrorOutput()->writeln(sprintf(
                'Published <info>%s</info> messages to queue <info>%s</info>',
                \count($messages),
                $queueName
            ));
        }

        return Command::SUCCESS;
    }

    private function readStdin(): string
    {
        $data = '';
        while (!feof(\STDIN)) {
            $data .= fread(\STDIN, 1024);
        }

        if ('' === $data) {
            throw new \InvalidArgumentException('No data received from STDIN');
        }

        return $data;
    }
}
